Include/internal/pycore_critical_section.h

```diff
@@ -98,6 +98,8 @@ extern "C" {
 # define Py_END_CRITICAL_SECTION()                                      \
         _PyCriticalSection_End(&_cs);                                   \
     }
+# define Py_EXIT_CRITICAL_SECTION()                                     \
```

Comment:
Maybe it is better to define a local API instead of a global one? For example `END_TYPE_LOCK()` does that.

Reply:
You can just extract the critical section into a helper function and then wrap the helper with the existing macros.

Reply:
For this API, I'd just extract most of `reversed_next` into a `reversed_next_impl` and wrap that with the `Py_{BEGIN,END}_CRITICAL_SECTION` macros. See also #120318.

Reply:
FWIW we ended up not doing that in #120318. There's [this comment](https://github.com/python/cpython/pull/120318#discussion_r1636802125) though that explains the approach and gives an example.

Reply:
Thanks for the feedback everyone! I added the `Py_EXIT_CRITICAL_SECTION` in this PR (and some others) to deal with return and goto statements. I am not in favor of adding a local API, since it will leak details of the locking implementation to  modules. Changing the `Py_EXIT_CRITICAL_SECTION` to a more private `_Py_EXIT_CRITICAL_SECTION` (or something like that) would be fine with me.

In this case we can indeed refactor to create a `reversed_next_impl` and put a global lock around it. I am not too worried about performance here, and if it turns out performance is an issue we can change later.

Unless more suggestions come it, I will refactor to `reversed_next_impl` in the coming days.